src/main/java/net/minecraftforge/common/BiomeDictionary.java

```diff
@@ -249,251 +220,231 @@ public static boolean areBiomesEquivalent(Biome biomeA, Biome biomeB)
         }
 
         if (biome.isHighHumidity() && biome.getBaseHeight() < 0.0F && (biome.getHeightVariation() <= 0.3F && biome.getHeightVariation() >= 0.0F))
         {
-            BiomeDictionary.registerBiomeType(biome, SWAMP);
+            BiomeDictionary.addTypes(biome, SWAMP);
         }
 
         if (biome.getBaseHeight() <= -0.5F)
         {
             if (biome.getHeightVariation() == 0.0F)
             {
-                BiomeDictionary.registerBiomeType(biome, RIVER);
+                BiomeDictionary.addTypes(biome, RIVER);
             }
             else
             {
-                BiomeDictionary.registerBiomeType(biome, OCEAN);
+                BiomeDictionary.addTypes(biome, OCEAN);
             }
         }
 
         if (biome.getHeightVariation() >= 0.4F && biome.getHeightVariation() < 1.5F)
         {
-            BiomeDictionary.registerBiomeType(biome, HILLS);
+            BiomeDictionary.addTypes(biome, HILLS);
         }
 
         if (biome.getHeightVariation() >= 1.5F)
         {
-            BiomeDictionary.registerBiomeType(biome, MOUNTAIN);
+            BiomeDictionary.addTypes(biome, MOUNTAIN);
         }
 
         if (biome.getEnableSnow())
         {
-            BiomeDictionary.registerBiomeType(biome, SNOWY);
+            BiomeDictionary.addTypes(biome, SNOWY);
         }
 
         if (biome.topBlock != Blocks.SAND && biome.getTemperature() >= 1.0f && biome.getRainfall() < 0.2f)
         {
-            BiomeDictionary.registerBiomeType(biome, SAVANNA);
+            BiomeDictionary.addTypes(biome, SAVANNA);
         }
 
         if (biome.topBlock == Blocks.SAND)
         {
-            BiomeDictionary.registerBiomeType(biome, SANDY);
+            BiomeDictionary.addTypes(biome, SANDY);
         }
         else if (biome.topBlock == Blocks.MYCELIUM)
         {
-            BiomeDictionary.registerBiomeType(biome, MUSHROOM);
+            BiomeDictionary.addTypes(biome, MUSHROOM);
         }
         if (biome.fillerBlock == Blocks.HARDENED_CLAY)
         {
-            BiomeDictionary.registerBiomeType(biome, MESA);
+            BiomeDictionary.addTypes(biome, MESA);
         }
     }
 
     //Internal implementation
     private static BiomeInfo getBiomeInfo(Biome biome)
     {
-        return biomeInfoMap.get(Biome.REGISTRY.getNameForObject(biome));
+        BiomeInfo info = biomeInfoMap.get(biome.getRegistryName());
+        if (info == null)
+        {
+            info = new BiomeInfo();
+            biomeInfoMap.put(biome.getRegistryName(), info);
+        }
+        return info;
     }
 
-    private static void checkRegistration(Biome biome)
+    /**
+     * Ensure that at least one type has been added to the given biome.
+     */
+    static void ensureHasTypes(Biome biome)
     {
-        if(!isBiomeRegistered(biome))
+        if (!hasAnyType(biome))
         {
+            FMLLog.warning("No types have been added to Biome %s, types will be assigned on a best-effort guess.");
```

Comment:
Alright I'm done trying to test this until you show me a screenshot in-game haha.
This line crashes.